File: rust/src/trace_context.rs
//! W3C Trace Context parsing for optional evidence ↔ external trace linking.
//! No OpenTelemetry SDK dependency — operators may attach IDs to evidence payloads.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ExternalTraceContext {
    pub trace_id: String,
    pub span_id: String,
    pub trace_flags: String,
}

/// Parse W3C `traceparent` (version 00): `00-<32hex>-<16hex>-<2hex>`.
pub fn parse_traceparent(value: &str) -> Option<ExternalTraceContext> {
    let s = value.trim();
    let parts: Vec<&str> = s.split('-').collect();
    if parts.len() != 4 || parts[0] != "00" {
        return None;
    }
    let trace_id = parts[1];
    let span_id = parts[2];
    let flags = parts[3];
    if trace_id.len() != 32
        || span_id.len() != 16
        || flags.len() != 2
        || !trace_id.chars().all(|c| c.is_ascii_hexdigit())
        || !span_id.chars().all(|c| c.is_ascii_hexdigit())
        || !flags.chars().all(|c| c.is_ascii_hexdigit())
    {
        return None;
    }
    Some(ExternalTraceContext {
        trace_id: trace_id.to_ascii_lowercase(),
        span_id: span_id.to_ascii_lowercase(),
        trace_flags: flags.to_ascii_lowercase(),
    })
}

/// JSON object suitable for optional `EvidenceEvent.payload["external_trace"]`.
pub fn external_trace_payload(ctx: &ExternalTraceContext) -> serde_json::Value {
    serde_json::json!({
        "trace_id": ctx.trace_id,
        "span_id": ctx.span_id,
        "trace_flags": ctx.trace_flags,
        "propagation": "w3c_traceparent",
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_valid_traceparent() {
        let tp = "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01";
        let ctx = parse_traceparent(tp).expect("parse");
        assert_eq!(ctx.trace_id.len(), 32);
        assert_eq!(ctx.span_id.len(), 16);
        assert_eq!(ctx.trace_flags, "01");
    }

    #[test]
    fn rejects_invalid_traceparent() {
        assert!(parse_traceparent("invalid").is_none());
    }
}

File: python/aigov_py/trace_context.py
"""W3C Trace Context helpers for linking GovAI evidence to external traces (no OTel SDK required)."""

from __future__ import annotations

import re
from typing import Any

_TRACEPARENT_RE = re.compile(
    r"^00-(?P<trace_id>[0-9a-fA-F]{32})-(?P<span_id>[0-9a-fA-F]{16})-(?P<flags>[0-9a-fA-F]{2})$"
)


def parse_traceparent(value: str | None) -> dict[str, str] | None:
    """Parse W3C ``traceparent`` header value into trace/span identifiers."""

    if not value or not str(value).strip():
        return None
    m = _TRACEPARENT_RE.match(str(value).strip())
    if not m:
        return None
    return {
        "trace_id": m.group("trace_id").lower(),
        "span_id": m.group("span_id").lower(),
        "trace_flags": m.group("flags").lower(),
        "propagation": "w3c_traceparent",
    }


def external_trace_payload(traceparent: str | None) -> dict[str, Any] | None:
    """Build optional ``payload.external_trace`` object for evidence ingest."""

    parsed = parse_traceparent(traceparent)
    if not parsed:
        return None
    return parsed


def attach_external_trace(payload: dict[str, Any], traceparent: str | None) -> dict[str, Any]:
    """Return a copy of ``payload`` with ``external_trace`` when ``traceparent`` is valid."""

    out = dict(payload)
    ext = external_trace_payload(traceparent)
    if ext:
        out["external_trace"] = ext
    return out
